### PR TITLE
Detect already-subscribed purchases in StoreKit 2

### PR DESCRIPTION
**Pull request by ajpallares**
### Checklist
- [x] Unit tests
- [ ] If applicable, create follow-up issues for `purchases-android` and hybrids

### Motivation

StoreKit 1 returns an `ASDServerError.currentlySubscribed` error when a user tries to purchase a subscription they already have. StoreKit 2, however, returns `.success` with the existing transaction — making it indistinguishable from a new purchase at the API level.

This means the SDK would re-post the receipt and the backend could create duplicate purchase events, which was identified while investigating inflated analytics for a customer using RevenueCat paywalls.

### Description

Before calling `Product.purchase()`, the SDK now captures the ID of the product's latest *finished* transaction. If the purchase returns the same transaction ID, it throws `productAlreadyPurchasedError`, aligning SK2 behavior with SK1.

The check is deliberately conservative — it only throws when **all** of these conditions are met:

1. The product has an existing transaction that has already been **finished** (unfinished transactions indicate a receipt that hasn't been posted yet, so retries must be allowed)
2. The current user already **owns this product** — either as an active subscription or a non-subscription purchase (so receipt transfers across RC account switches and expired subscription re-purchases are not blocked)
3. The purchase is **not** using a promotional or win-back offer (these are explicitly for existing subscribers)
4. The purchase returned the **same transaction ID** as the pre-existing one

This is implemented in `PurchasesOrchestrator` via a new `finishedTransactionID(for:)` helper that checks `Product.latestTransaction` against `Transaction.unfinished` and the cached `CustomerInfo`.

Edge cases that correctly allow the purchase through:
- **Unfinished transactions**: receipt may not have been posted (e.g., server was down)
- **Different RC user on same Apple ID**: receipt needs to be posted for the new user
- **Expired subscriptions**: SK2 creates a new transaction on re-subscribe
- **Promotional/win-back offers**: these are for existing subscribers by definition
- **Consumable re-purchases**: SK2 creates a new transaction ID each time
- **No cached customer info (cold start)**: can't determine ownership, fails open

---

> [!NOTE]
> **Medium Risk**
> Changes StoreKit 2 purchase behavior by throwing `productAlreadyPurchasedError` when SK2 returns an existing transaction, which could affect purchase flows and error handling in production if the detection misfires. Risk is mitigated by conservative checks and extensive unit test coverage for edge cases (offers, unfinished transactions, account switches, expired subs).
> 
> **Overview**
> Prevents duplicate receipt posts / purchase events in StoreKit 2 by detecting when `Product.purchase()` returns an *existing* transaction for an already-owned product and surfacing this as `productAlreadyPurchasedError`.
> 
> Adds `SK2AlreadySubscribedDetector` and integrates it into `PurchasesOrchestrator` (skipping the check for promotional/win-back/custom offer purchases), plus a new log message for the detected condition. Expands SK2 unit tests to cover the new detection logic and adjusts existing tests (e.g., clearing StoreKit test-session transactions) to avoid false positives.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit f3cb3e4b8029cdcc3804124eebbff3b8f7bb19c5. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: Sources/Purchasing/Purchases/PurchasesOrchestrator.swift [hunk 1]
```diff
@@ -764,6 +764,13 @@ final class PurchasesOrchestrator {
 
             self.cachePresentedOfferingContext(package: package, productIdentifier: sk2Product.id)
 
+            // Promotional and win-back offers are for existing subscribers,
+            // so SK2 returning the same transaction is expected — skip the check.
```
Comment (tonidero) [inline Sources/Purchasing/Purchases/PurchasesOrchestrator.swift]: `skip the check` probably won't mean anything to us down the line... I would say we might want to add more details of what this is doing?

File: Sources/Purchasing/Purchases/PurchasesOrchestrator.swift [hunk 2]
```diff
@@ -764,6 +764,13 @@ final class PurchasesOrchestrator {
...
+            let isOfferPurchase = promotionalOffer != nil || winBackOffer != nil
+            let existingTransactionID = isOfferPurchase
+                ? nil
+                : await self.finishedTransactionID(for: sk2Product)
```
[the remark this answers is not in the document]
Comment (tonidero) [reply Sources/Purchasing/Purchases/PurchasesOrchestrator.swift]: Hmm this seems like it could be valid?

File: Sources/Purchasing/Purchases/PurchasesOrchestrator.swift [hunk 3]
```diff
@@ -886,6 +902,43 @@ final class PurchasesOrchestrator {
...
 
+    /// Returns the transaction ID of the product's latest transaction only when the product
+    /// has already been fully processed for the current user.
+    ///
+    /// Returns `nil` (allowing the purchase through) when:
+    /// - There is no existing transaction
+    /// - The existing transaction is unfinished (receipt may not have been posted yet)
+    /// - The current user hasn't purchased this product (e.g., after switching RC accounts)
+    @available(iOS 15.0, tvOS 15.0, watchOS 8.0, macOS 12.0, *)
+    private func finishedTransactionID(for product: SK2Product) async -> String? {
```
Comment (tonidero) [inline Sources/Purchasing/Purchases/PurchasesOrchestrator.swift]: I do wonder if there is a chance to extract most of this logic to a helper... Again, mostly trying to keep `PurchasesOrchestrator` from growing more 😅

File: Sources/Purchasing/Purchases/PurchasesOrchestrator.swift [hunk 4]
```diff
@@ -886,6 +902,43 @@ final class PurchasesOrchestrator {
         #endif
     }
 
+    /// Returns the transaction ID of the product's latest transaction only when the product
+    /// has already been fully processed for the current user.
+    ///
+    /// Returns `nil` (allowing the purchase through) when:
+    /// - There is no existing transaction
+    /// - The existing transaction is unfinished (receipt may not have been posted yet)
+    /// - The current user hasn't purchased this product (e.g., after switching RC accounts)
```
Comment (tonidero) [inline Sources/Purchasing/Purchases/PurchasesOrchestrator.swift]: I do wonder that these docs do not reflect the name of this method IMO... Maybe we should make the method name more specific to this use case somehow? Maybe like? Maybe like `alreadyPurchasedTransactionID`?

**Review by tonidero — COMMENTED** [review]
This is looking great to me, but would like catforms to take a look 🙏 Just a couple comments but nothing blocking for me.
